rts/motoko-rts/src/gc/incremental/sort.rs

```diff
@@ -0,0 +1,41 @@
+use core::cmp::Ordering;
+
+/// Simple in-place sort algorithm that does not allocate additional memory.
+/// Based on quicksort. Average runtime costs of `O(n*log(n))`.
+pub fn sort<Compare: Fn(usize, usize) -> Ordering>(array: &mut [usize], compare: &Compare) {
+    if array.len() > 1 {
+        quicksort(array, compare);
+    }
+}
+
+fn quicksort<Compare: Fn(usize, usize) -> Ordering>(array: &mut [usize], compare: &Compare) {
+    debug_assert!(array.len() > 1);
+    let pivot = array[array.len() / 2];
+    let mut left = 0;
+    let mut right = array.len() - 1;
+    loop {
+        while compare(array[left], pivot) == Ordering::Less {
+            left += 1;
+        }
+        while compare(array[right], pivot) == Ordering::Greater {
+            right -= 1;
+        }
+        if left <= right {
+            // Cannot use `array.swap()` since it imports `memmove` in debug build.
+            let temporary = array[left];
+            array[left] = array[right];
+            array[right] = temporary;
+            left += 1;
+            right -= 1;
+        }
+        if left > right {
+            break;
+        }
+    }
+    if right > 0 {
+        quicksort(&mut array[..right + 1], compare);
+    }
+    if left < array.len() - 1 {
+        quicksort(&mut array[left..], compare);
+    }
```

Comment:
```suggestion
    if right < array.len() - 1 {
        quicksort(&mut array[right + 1..], compare);
    }
    if left > 0  {
        quicksort(&mut array[..left], compare);
    }
```
These recursive calls look wrong to me, but perhaps I'm misunderstanding Rust or this version of quicksort. Shouldn't it be something like the above?

Reply:
Interesting point: It is indeed a special version I have adopted from Niklaus Wirth (https://people.inf.ethz.ch/wirth/AD.pdf, pages 65/66): After the partitioning, the invariant is `left > right`. The original version uses variable names `i` and `j`. I will add a comment for clarification. Maybe, I should also choose better variable names instead?
Side note: I noticed that the examples in our samples could run in endless recursion if the pivot would not be chosen as the first element (which is bad for sorted or nearly sorted arrays, regardless of the order). Here, the pivot is the middle element which is good for (nearly) sorted arrays, which could be helpful in scenarios that follow the generational hypothesis where low-garbage old generations cold be populated in linear fashion. Although, one can always create opposite examples...